system/common.cpp

```diff
@@ -205,5 +210,114 @@ int WEAK pthread_cancel(struct __pthread* t)
 	return __syscall_exit(EX_SOFTWARE);
 }
 
+#  define REPLACEMENT_CHARACTER 0xFFFD
+#  define MAX_CODEPOINT 0x10FFFF
+#  define INVALID_CODEPOINT -1
+static size_t utf16_to_utf8(char *dest, size_t dlen, const char16_t *src, size_t slen) {
+  size_t j = 0;
```

Comment:
two questions:
- don't we have something in client::String to convert to utf8?
- either way, isn't it better to convert from client::String directly to utf8 instead of using the intermediate char16_t buffer?

Reply:
1. In theory yes, but it assumes ASCII, thus unicode will not work
2. I'll try it out, the main reason why I did it like this is because it's the same as I did in ASan

Reply:
@Hyxogen I thought that Chen added the toUtf8 method, but actually it's only in the new clientlib.
You could copy it from src/extensions.ts in ts2cpp, and then we will just use that when the new clientlib is integrated

Reply:
actually, you will need to customize it anyway, because you want a fixed size buffer, not a std::string here.